File: packages/mcp/src/ndjson.ts
export const splitNdjson = (buffer: string, chunk: string): {
  readonly lines: ReadonlyArray<string>
  readonly rest: string
} => {
  const combined = buffer + chunk
  const parts = combined.split("\n")
  const rest = parts.pop() ?? ""
  return {
    lines: parts.filter((line) => line.trim().length > 0),
    rest,
  }
}

export const encodeNdjson = (value: unknown): string => `${JSON.stringify(value)}\n`

File: beard/core/src/main/scala/groksbeard/core/Ndjson.scala
package groksbeard.core

object Ndjson:
  def split(buffer: String, chunk: String): (List[String], String) =
    val combined = buffer + chunk
    val parts    = combined.split("\n", -1).toList
    val rest     = parts.lastOption.getOrElse("")
    val lines    = parts.dropRight(1).map(_.trim).filter(_.nonEmpty)
    (lines, rest)

  def encode(jsonLine: String): String =
    jsonLine + "\n"

  def encodeChunk(jsonLines: List[String]): String =
    if jsonLines.isEmpty then "" else jsonLines.mkString("\n") + "\n"
end Ndjson
